### PR TITLE
Remove function

### PR DESCRIPTION
**Pull request by Dor8**

### CODE REVIEW

File: WireBST.java [hunk 1]
```diff
@@ -97,20 +97,94 @@ private void insertToRight(WireBSTNode current , WireBSTNode newNode ){
...
+            }
+        }
+    }
 
+    /**
+     * removing a deep leaf. used by the removeWiredBSTNode function
+     * @param node. the node we wish to remove
+     */
+    private void removeDeepLeaf( WireBSTNode node ){
+        if ( node.getParent() = null ){
```
Comment (reahaas) [inline WireBST.java]: need to be ==

File: WireBST.java [hunk 2]
```diff
@@ -97,20 +97,94 @@ private void insertToRight(WireBSTNode current , WireBSTNode newNode ){
...
+    /**
+     * remove of the node WireBST
+     * @param node is the node to remove.
+     */
+    public void removeWireBSTNode(WireBSTNode node){
+        if ( node == null) return;
+        if ( ! node.isRealLeft() && ! node.isRealRight() ){
+            removeDeepLeaf(node);
+        }
+        if ((node.getRight() != null) && (node.getleft() == null)){    //** in case of removing node who has only right son
```
Comment (reahaas) [inline WireBST.java]: need to be: if ( node.isRealRight() && (! node.isRealLeft() ) )

no null in the tree except min & max :)

File: WireBST.java [hunk 3]
```diff
@@ -97,20 +97,94 @@ private void insertToRight(WireBSTNode current , WireBSTNode newNode ){
...
+            removeNodeWithOnlyRightSon(node);
+        }
```
[the remark this answers is not in the document]
Comment (reahaas) [reply WireBST.java]: לא הערה מספיק נכונה, צריך לחשוב פה
this.setHead(node.getRight());
this.getHead().setLeft(null);
this.getHead().setParent(null);

File: WireBST.java [hunk 4]
```diff
@@ -149,21 +232,54 @@ public void maxWireBST(WireBSTNode node){
...
+        preOrderScan(node.getRight());                          //recursive call on right subtree
 
     }
 
-    // print the WireBST as in order
+    /**
+     * print the wireBST as in order
+     * @param node
+     */
     public void inOrderScan(WireBSTNode node){
```
Comment (reahaas) [inline WireBST.java]: better be like this:
WireBSTNode temp = minWireBST() ;   //  temp now poit on the minimun of the tree, which is the leftest node.
for ( ; temp; temp = temp.getLeft() ) 
     toString(temp);                           // toString will return a string the describe the node (without the wires) : " <studentNum>: <studentName> "

File: WireBST.java [hunk 5]
```diff
@@ -149,21 +232,54 @@ public void maxWireBST(WireBSTNode node){
...
+     */
     public void postOrderScan(WireBSTNode node){
 
+        if (node == null) return;
+
+        postOrderScan(node.getLeft());                          //first, recursive call on left subtree
+
+        postOrderScan(node.getRight());                         //then recursive call on right subtree
+
+        System.out.println(node._studentNum + " ");             //last, print node data
```
Comment (reahaas) [inline WireBST.java]: make a function:
printNodeData()
!!!

File: src/WireBST.java [hunk 6]
```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
...
             }
+        }
+        else if (this.getNodeCounter() % 2 == 1) {
+            this.setMedian(getPredecessor(this.getMedianWireBST()));
+        }
         this._nodeCounter ++;
+    }
+
+    /**
+     * remove the student that is number is the given parameter
```
Comment (Dor8) [inline src/WireBST.java]: his

File: src/WireBST.java [hunk 7]
```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
...
+     * @param studentNum, the node who include that number
```
Comment (reahaas) [inline src/WireBST.java]: key for node to remove

File: src/WireBST.java [hunk 8]
```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
...
+        }
+    }
+
+    /**
+     * overloading the removeWireBSTNode function, remove node who exist in the tree
+     * @param node is the node to remove.
+     */
+    public void removeWireBSTNode(WireBSTNode node){
+
+        if ( ! node.isRealLeft() && ! node.isRealRight() ){
```
Comment (reahaas) [inline src/WireBST.java]: change to : 
if()
else if()
else

File: src/WireBST.java [hunk 9]
```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
...
+    private void removeDeepLeaf(WireBSTNode node){
+
+        if (node.getParent() == null){                              //in case node is root and has no sons
+
+            this.setHead(null);
+            this.setMedian(null);
+
+        }
+
+        else if (node.equal(node.getParent().getRight())) {
```
Comment (reahaas) [inline src/WireBST.java]: // case node is a right son

File: src/WireBST.java [hunk 10]
```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
...
+
+            updateMedianRemove(node);
+            node.getParent().setRight(getSuccessor(node));
+        }
+
+        else {
```
Comment (reahaas) [inline src/WireBST.java]: // case node is a left son

File: src/WireBST.java [hunk 11]
```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
...
+        if (node.getParent() == null){
+
+            updateMedianRemove(node);
+            node.getRight().setParent(node.getParent());
+            this.setHead(node.getRight());
+            getSuccessor(node).setLeft(node.getLeft());
+
+        }
+
+        else if  (node.equal(node.getParent().getRight())){
```
Comment (reahaas) [inline src/WireBST.java]: // case this node is a right son

File: src/WireBST.java [hunk 12]
```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
...
+
+        }
+    }
+
+    /**
+     * update the median after remove of a node
+     * @param node, the node that removed
+     */
+    private void updateMedianRemove(WireBSTNode node){
+
```
Comment (reahaas) [inline src/WireBST.java]: if( this.getCounter() == 1 )
this.setMedian(null);

File: src/WireBST.java [hunk 13]
```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
...
+    /**
+     * removing a deep leaf. used by the removeWiredBSTNode function
+     * @param node the node we wish to remove
+     */
+    private void removeDeepLeaf(WireBSTNode node){
+
+        if (node.getParent() == null){                              //in case node is root and has no sons
+
+            this.setHead(null);
+            this.setMedian(null);
```
Comment (reahaas) [inline src/WireBST.java]: remove this line

File: src/WireBST.java [hunk 14]
```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
...
+
+        WireBSTNode temp = searchWireBST( studentNum );
+
+        if (temp == null){
+
+            return false;
+        }
+
+        else {
+
```
Comment (reahaas) [inline src/WireBST.java]: need to add:
updateMedianRemove(node);

Comment (reahaas) [reply src/WireBST.java]: and remove all the other duplicated sub-calls to this in the sub-functions of remove

File: src/WireBST.java [hunk 15]
```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
...
+    }
+
+    /**
+     * removing a node with only right son
+     * @param node the node we wish to remove
+     */
+    private void removeNodeWithOnlyRightSon(WireBSTNode node){
+
+        if (node.getParent() == null){
```
Comment (reahaas) [inline src/WireBST.java]: // case node is root

File: src/WireBST.java [hunk 16]
```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
...
+        else if  (node.equal(node.getParent().getRight())){
+
+            updateMedianRemove(node);
+            node.getRight().setParent(node.getParent());
+            node.getParent().setRight(node.getRight());
+            getSuccessor(node).setLeft(node.getParent());
+
+        }
+
+        else {
```
Comment (reahaas) [inline src/WireBST.java]: // case this node is a left son

File: src/WireBST.java [hunk 17]
```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
...
+        }
+    }
+
+
+
+    /**
+    * removing a node with only left son
+    * @param node the node we wish to remove
+    */
+    private void removeNodeWithOnlyLeftSon(WireBSTNode node){
```
Comment (reahaas) [inline src/WireBST.java]: add documentation to cases

File: src/WireBST.java [hunk 18]
```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
...
+
+        WireBSTNode temp = getSuccessor(node);
+
+        if (node.getParent() == null ){
+
+            this.setHead(temp);
+            temp.setParent(node.getParent());
+            temp.setRight(node.getRight());
+            node.getRight().setParent(temp);
+            temp.setLeft(node.getLeft());
```
Comment (reahaas) [inline src/WireBST.java]: temp.getParent().setLeft(temp.getRight());

Comment (reahaas) [reply src/WireBST.java]: // update the parent of the temp, to point on the right son of temp (temp cant have a left son, cause he is the successor of the nodeToRemove)

File: src/WireBST.java (hunk 7 above)
Comment (Dor8) [reply src/WireBST.java]: done

File: src/WireBST.java (hunk 13 above)
Comment (Dor8) [reply src/WireBST.java]: done

File: src/WireBST.java (hunk 9 above)
Comment (Dor8) [reply src/WireBST.java]: done

File: src/WireBST.java (hunk 10 above)
Comment (Dor8) [reply src/WireBST.java]: done

File: src/WireBST.java (hunk 12 above)
Comment (Dor8) [reply src/WireBST.java]: done

File: src/WireBST.java (hunk 15 above)
Comment (Dor8) [reply src/WireBST.java]: done

File: src/WireBST.java (hunk 11 above)
Comment (Dor8) [reply src/WireBST.java]: done

**Pull request merged**